src/sfctl/custom_app_type.py

```diff
@@ -6,53 +6,108 @@
+       import (ProvisionApplicationTypeDescription)
+    from azure.servicefabric.models.external_store_provision_application_type_description \
+        import (ExternalStoreProvisionApplicationTypeDescription)
+
+    from azure.servicefabric.models import FabricErrorException
+
+    provision_description = None
+
+    # Validate inputs
+    if not external_provision:
+        if not application_type_build_path:
+            raise CLIError('Missing required parameter '
+                           '--application-type-build-path.')
+
+        provision_description = \
+            ProvisionApplicationTypeDescription(
+                no_wait,
+                application_type_build_path=application_type_build_path
+                )
     else:
-        provision_desc = ProvisionApplicationTypeDescription(
-            no_wait,
-            application_type_build_path=application_type_build_path
-        )
+        if not all([application_package_download_uri, application_type_name, \
+            application_type_version]):
+            raise CLIError('Missing required parameters. The following are required: '
+                           '--application-package-download-uri, --application-type-name, '
+                           '--application-type-version.')
+        provision_description = \
+            ExternalStoreProvisionApplicationTypeDescription(
+                no_wait,
+                application_package_download_uri=\
+                    application_package_download_uri,
+                application_type_name=application_type_name,
+                application_type_version=application_type_version)
+
+    api_version = "6.1"
+
+    # Construct URLs
+    url = '/ApplicationTypes/$/Provision'
+
+    # Construct parameters
+    query_parameters = {}
+    query_parameters['api-version'] = client._serialize.query(
+        "api_version", api_version, 'str')
+
+    query_parameters['timeout'] = client._serialize.query(
+        "timeout",
+        timeout,
+        'long',
+        maximum=4294967295,
+        minimum=1)
+
+    # Construct headers
+    header_parameters = {}
+    header_parameters['Content-Type'] = 'application/json; charset=utf-8'
+
+    # Construct body
+    body_content = None
+    if not external_provision:
+        body_content = client._serialize.body(
+            provision_description,
+            'ProvisionApplicationTypeDescription')
+    else:
+        body_content = client._serialize.body(
+            provision_description,
+            'ExternalStoreProvisionApplicationTypeDescription')
+
+    # Create a new sorted dictionary since we don't have move_to_end in python 2
+    body_content_sorted = OrderedDict([('Kind', body_content['Kind'])])
+    for key in body_content:
+        if key != 'Kind':
+            body_content_sorted[key] = body_content[key]
+
+    if list(body_content_sorted.keys())[0] != "Kind":
+        raise CLIError('Internal CLI error: Kind must be the first item to be serialized.')
```

Comment:
If this an internal error, then it shouldn't be a `CLIError`, it should be a different kind. The type `CLIError` is meant to be user-actionable, like a wrong parameter or something else the user can immediately fix.